### PR TITLE
Add multimodal sequence classification support for Mistral 3

### LINKED ISSUE #48187: Add multimodal sequence classification support for Mistral 3
### Feature request

Add `Mistral3ForSequenceClassification` using the existing multimodal backbone. The proposed class would:

- work with `AutoModelForSequenceClassification`
- support image, text, and combined image-text inputs
- support `bos`, `eos`, and masked-mean pooling, with `eos` as the default
- allow `text_config.is_causal=False` for bidirectional reranking while preserving causal attention by default

### Motivation

Sequence classification is supported for the Ministral 3 text model, but not for the full Mistral 3 vision-language model. This means image-text rerankers currently need a custom model class and exported checkpoints require `trust_remote_code=True`.

### Your contribution

I have a working implementation here:

#48189

It follows the existing multimodal classification pattern.

### PR DESCRIPTION
**Pull request by oliverholworthy**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48189) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48189)

# What does this PR do?

Adds native `Mistral3ForSequenceClassification` support using the existing multimodal backbone. This allows image-text reranker checkpoints to load through `AutoModelForSequenceClassification` without custom code or `trust_remote_code=True`.

The classifier supports `bos`, `eos`, and masked-mean pooling, following the recently added [Cohere Compass classifier](https://github.com/huggingface/transformers/blob/f93eb6dbec8c2f5563f6d69bd998771929e8263f/src/transformers/models/cohere_compass/modular_cohere_compass.py#L509-L593) (#47878). A possible follow-up might be to consolidate some of these pooling options into `GenericForSequenceClassification`, but that broader change is outside this PR.

The change also allows an optional `text_config.is_causal=False` value to reach Ministral 3 attention while preserving causal attention by default.

Tests cover pooling, bidirectional attention, multimodal inputs, and save/reload through the Auto class. The relevant model and repository consistency checks pass.

Fixes #48187

## Who can review?

@zucchini-nlp

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
I think we're not adding a SeqClf class anymore unless there is clearly a need (e.g. an official ckpt released by Cohere team)

I will close the issue and the PR

**Comment by oliverholworthy** [comment]
@zucchini-nlp  Thanks for the review. One piece of context I should have included in the motivation is that this is driven by a concrete checkpoint requirement rather than API completeness. I’m part of NVIDIA’s retrieval model team, which works on models such as [Nemotron 3 Embed](https://huggingface.co/nvidia/Nemotron-3-Embed-1B-BF16).

Would you be open to discussing whether the use case meets the bar for a model-specific class?

File: src/transformers/models/mistral3/modeling_mistral3.py [hunk 1]
```diff
@@ -322,6 +328,105 @@ def forward(
...
+    custom_intro="""
+    The Mistral3 model with a sequence classification head on top.
+    """
+)
+class Mistral3ForSequenceClassification(GenericForSequenceClassification, Mistral3PreTrainedModel):
+    r"""Sequence classification head.
+
+    - ``eos`` (default when ``pooling`` is ``None``): the rightmost non-pad token.
+    - ``bos``: the first token. Meaningful only when the text backbone is bidirectional.
+    - ``mean``: masked mean over non-pad hidden states, followed by the score layer.
```
Comment (zucchini-nlp) [inline src/transformers/models/mistral3/modeling_mistral3.py:341]: do we need all of them, i,e, released models has different values across ckpts?

File: src/transformers/models/ministral3/modular_ministral3.py [hunk 2]
```diff
@@ -32,6 +33,10 @@ def get_llama_4_attn_scale(positions_ids: torch.Tensor, beta: float, max_positio
 
 
 class Ministral3Attention(MistralAttention):
+    def __init__(self, config: Ministral3Config, layer_idx: int):
+        super().__init__(config, layer_idx)
+        self.is_causal = getattr(config, "is_causal", True)
+
```
Comment (zucchini-nlp) [inline src/transformers/models/ministral3/modular_ministral3.py]: just to be sure this is intended, the classifier model uses a bidirectional attn in LM backbone right?

File: tests/models/mistral3/test_modeling_mistral3.py [hunk 3]
```diff
@@ -193,6 +199,89 @@ def check_config_can_be_init_without_params():
...
+        self.assertTrue(torch.isfinite(output.logits).all())
+        torch.testing.assert_close(output.logits, torch.zeros_like(output.logits))
+
+    def test_sequence_classification_multimodal_auto_round_trip(self):
+        config, inputs = self.model_tester.prepare_config_and_inputs_for_common()
+        config.num_labels = 1
+        config.pooling = "mean"
+        config.text_config.is_causal = False
+        model = Mistral3ForSequenceClassification(config).to(torch_device).eval()
+
```
Comment (zucchini-nlp) [inline tests/models/mistral3/test_modeling_mistral3.py:260]: looks kinda same as `test_save_load` from base class, might be redundant

File: tests/models/mistral3/test_modeling_mistral3.py [hunk 4]
```diff
@@ -193,6 +199,89 @@ def check_config_can_be_init_without_params():
...
+            model.model = StaticBackbone(hidden_states)
+            with torch.no_grad():
+                model.score.weight.zero_()
+                model.score.weight[0, 0] = 1
+                output = model(input_ids=input_ids, attention_mask=attention_mask)
+            expected = torch.tensor([[expected_score]], device=torch_device)
+            torch.testing.assert_close(output.logits, expected)
+
+    def test_sequence_classification_mean_pooling_precedes_score_projection(self):
+        from transformers.modeling_outputs import BaseModelOutputWithPast
```
Comment (zucchini-nlp) [inline tests/models/mistral3/test_modeling_mistral3.py]: lets import at the top of the file, once

**Review by zucchini-nlp — COMMENTED** [review]
Sorry, forgot to reopen it yesterday! 

I am not sure if we really need all 3 pooling options, we usually don't add code branching unless the released ckpt actually needs it. Can you confirm that the ckpts have all types of pooling?

Also cc @tomaarsen , just to make sure we're hosting these on `transformers` side since v6 added a lot of new feats

File: src/transformers/models/mistral3/modeling_mistral3.py (hunk 1 above)
Comment (oliverholworthy) [reply src/transformers/models/mistral3/modeling_mistral3.py:341]: We don’t need all three. The concrete checkpoint work currently needs mean pooling and last-token pooling while we evaluate those two options  (mean is the most likely outcome). I can remove bos and use the Sentence Transformers terminology `mean` and `lasttoken`. The last-token path is already the GenericForSequenceClassification behavior, so the only custom pooling implementation would be `mean` pooling.

File: src/transformers/models/ministral3/modular_ministral3.py (hunk 2 above)
Comment (oliverholworthy) [reply src/transformers/models/ministral3/modular_ministral3.py]: This was intended, due to bidirectional attention requirement. However, on a second look it seems the model-forward path already propagates config.is_causal to masking and the attention backend, so changing the attention module itself may not strictly be needed. Will investigate separately to be sure.
Removed from this PR.

File: tests/models/mistral3/test_modeling_mistral3.py (hunk 4 above)
Comment (oliverholworthy) [reply tests/models/mistral3/test_modeling_mistral3.py]: thanks for spotting this, import moved to top

File: tests/models/mistral3/test_modeling_mistral3.py (hunk 3 above)
Comment (oliverholworthy) [reply tests/models/mistral3/test_modeling_mistral3.py:260]: there is some redundancy here, the main benfit of this additional test is to check that the class specific config like the pooling in particular is restored correctly from round trip. I can narrow the test a bit or remove it if you think that's sufficiently captured by other tests or base class

File: src/transformers/models/ministral3/modular_ministral3.py (hunk 2 above)
Comment (tomaarsen) [reply src/transformers/models/ministral3/modular_ministral3.py]: Yes, `is_causal` is propagated quietly through one of the decorators, allowing any causal architecture to become bidirectional at will. It's very useful for models in information retrieval.

**Comment by zucchini-nlp** [comment]
@oliverholworthy We discussed this internally and found that model can be easily hosted in Sentence Transformers by re-using a `Mistral3Model`. Sentence Transformers has a way to add a necessary pooling and linear layers on top, Tom will guide you through

LMK if that sounds good to you

**Comment by tomaarsen** [comment]
Hello!

@oliverholworthy As you're well aware, `transformers` ships both model "bodies" (i.e. via `AutoModel`) and "heads" (i.e. via `AutoModelForCausalLM`, `AutoModelForSequenceClassification`, etc.), where the bodies do the bulk of the work and the heads are mostly for converting whatever the body created to the correct output contract. For transformers, it's crucial to maintain the bodies, but some of the heads can also be deferred to other projects that define their own output contract. I think that's the case here too, where Sentence Transformers already has the Modules in place for what you're looking for, while also being more flexible. 

For example, via Sentence Transformers, you could use input unpadding for text models with FA2, efficiently process the whole batch, and then go back to normal per-item Pooling. Via ST you also have more options in terms of pooling, for example.

As of the Sentence Transformers v5.4 release, which added multimodality support to ST & refactored its `CrossEncoder` class, you can create `CrossEncoder` instances in many more ways than just a `transformers` `...ForSequenceClassification`. For example, the [Qwen3-VL-Reranker](https://huggingface.co/Qwen/Qwen3-VL-Reranker-2B) models use [`"any-to-any"`](https://huggingface.co/Qwen/Qwen3-VL-Reranker-2B/blob/main/sentence_bert_config.json#L2) (which wraps `AutoModelForMultimodalLM`) to generate a "next token" given the input, and a [`LogitScore` module](https://huggingface.co/Qwen/Qwen3-VL-Reranker-2B/blob/main/1_LogitScore/config.json) to read the difference between the "yes" and "no" tokens.

Or the [nvidia/llama-nemotron-rerank-vl-1b-v2](https://huggingface.co/nvidia/llama-nemotron-rerank-vl-1b-v2) model, which uses custom code to wrap `AutoModelForSequenceClassification`, which ST can load via [`"sequence-classification"`](https://huggingface.co/nvidia/llama-nemotron-rerank-vl-1b-v2/blob/main/sentence_bert_config.json#L2). 

But in that case, it's also possible to instead use `"transformers_task": "feature-extraction"` (meaning `AutoModel`), and use a Pooling module to pool using whatever strategy is desired, plus a Dense module to apply whatever Linear they want. Then the model loads fully out-of-the-box using stock transformers and stock Sentence Transformers, with native multimodality and no `trust_remote_code` or any risk of custom code going stale.

And if custom code is still required due to some unusual changes that might not fit natively in either transformers or ST (e.g. let's say attention-based pooling or something), then you can still write custom code for either `transformers` (like that last model) or for ST directly, e.g. like https://huggingface.co/perplexity-ai/pplx-embed-v1-0.6b/blob/main/st_quantize.py, which adds a Module that automatically and always quantizes the outputs to int8/binary.

I think as far as rerankers go: there's a lot of usage via ST, so you can take advantage of that nicely. Let me know if you have any issues with that proposal, though.

- Tom Aarsen

**Comment by oliverholworthy** [comment]
Thanks for the detailed reply @tomaarsen , this all makes sense. I’ve validated that the Feature Extraction + Pooling + Dense can represent the intended multimodal reranker and preserve the trained scoring weight. I agree that this is more flexible than adding new model-specific pooling options to Transformers, and enables us to experiment with different designs without depending on an updates to transformers.

We’ll look into leveraging the Sentence Transformers layout and handle training/export integration. The remaining interoperability gap for us is vLLM, which does not yet interpret this module stack as a CrossEncoder or fully preserve the structured multimodal pair input, so will follow that up separately in vLLM.

**Comment by oliverholworthy** [comment]
Closing this since Sentence Transformers covers the intended use-case here

**Comment by tomaarsen** [comment]
I'm glad to hear that it works for you! I'm not sure if you're in any joint NVIDIA <> HF Slack channels, but if yes, you can also reach me there, or via LinkedIn, etc. I'd be happy to help with any integration issues or with sharing the word on the releases once they're out.

- Tom Aarsen
